Bearded.Utilities/Algorithms/CoffmanGraham.cs

```diff
@@ -0,0 +1,196 @@
+                    // of sources in a directed acyclic graph where all already elements and adjacent arrows have been
+                    // removed.
+                    var sources = elements.Where(allPredecessorsHaveBeenOrdered);
+
+                    // For each considered element, we look at the place of all their predecessors in the partial
+                    // topological ordering we have constructed so far.
+                    // * We first consider the most recently added predecessor of each element. That is, the predecessor
+                    //   of said element that is the highest in the current partial topological ordering.
+                    // * Of all these predecessors, we pick the predecessor that comes earliest in the ordering. If
+                    //   there is only one element with said predecessor, that element is added to the ordering next.
+                    // * Ties are broken by looking at the next highest ordered predecessor of all tied elements.
+                    // * If all predecessors of 2 or more elements are the same, we pick the element with the least
+                    //   predecessors.
+                    // This selection process is implemented by representing the predecessors of an element as a
+                    // decreasing number sequence of the predecessors' indices, and selecting the lowest of those in a
+                    // reflected lexicographic ordering.
+                    var next = sources.MinBy(createDecreasingNumberSequenceOfPredecessorIndices);
+
+                    orderedElementIndices.Add(next, ordering.Count);
+                    ordering.Add(next);
+                    elements.Remove(next);
+                }
+
+                return ordering;
+
+                bool allPredecessorsHaveBeenOrdered(T e) => graph.GetDirectPredecessorsOf(e)
+                    .All(n => orderedElementIndices.ContainsKey(n));
+
+                DecreasingNumberSequence createDecreasingNumberSequenceOfPredecessorIndices(T e)
+                {
+                    var predecessorIndices = graph.GetDirectPredecessorsOf(e).Select(n => orderedElementIndices[n]);
+                    return DecreasingNumberSequence.FromUnsortedNumbers(predecessorIndices);
+                }
+            }
+
+            private static ImmutableList<ImmutableHashSet<T>> createLayers(
+                // ReSharper disable once SuggestBaseTypeForParameter
+                IDirectedAcyclicGraph<T> graph, IList<T> ordering, int maxLayerSize)
+            {
+                var layersReversed = new List<ImmutableHashSet<T>.Builder>();
+                var elementToLayer = new Dictionary<T, int>();
+
+                for (var i = ordering.Count - 1; i >= 0; i--)
+                {
+                    // We fill in the layers back-to-front, always choosing the highest layer for the current element,
```

Comment:
So far you're talking about high/low. The back/front here confuses me a little. I think I understand it (we are building the resulting list in reverse order starting with the last layer, right?) but I'm not 100% sure.

Also, maybe you can mention why you do this? Why wouldn't front-to-back work?

Reply:
I really don't think we should aim to prove or explain the correctness of the algorithm here. The file is already comment-heavy, and Wikipedia is only one click away.